prov/rxm/src/rxm_ep.c

```diff
@@ -380,17 +380,96 @@ static int rxm_getname(fid_t fid, void *addr, size_t *addrlen)
 	return fi_getname(&rxm_ep->msg_pep->fid, addr, addrlen);
 }
 
+static int rxm_mc_close(struct fid *fid)
+{
+	struct rxm_mc *rxm_mc;
+
+	rxm_mc = container_of(fid, struct rxm_mc, mc_fid.fid);
+
+	ofi_atomic_dec32(&rxm_mc->av_set->ref);
+
+	if (rxm_mc->util_coll_mc_fid) {
+		assert (rxm_mc->util_coll_join_completed == 1);
+		fi_close(&rxm_mc->util_coll_mc_fid->fid);
+	}
+
+	if (rxm_mc->offload_coll_mc_fid) {
+		assert (rxm_mc->offload_coll_join_completed == 1);
+		fi_close(&rxm_mc->offload_coll_mc_fid->fid);
+	}
+
+	free(rxm_mc);
+
+	return FI_SUCCESS;
+}
+
+static struct fi_ops rxm_mc_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = rxm_mc_close,
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static struct rxm_mc *rxm_create_mc(struct util_av_set *av_set,
+					void *context)
+{
+	struct rxm_mc *rxm_mc;
+
+	rxm_mc = calloc(1, sizeof(*rxm_mc));
+	if (!rxm_mc)
+		return NULL;
+
+	rxm_mc->mc_fid.fid.fclass = FI_CLASS_MC;
+	rxm_mc->mc_fid.fid.context = context;
+	rxm_mc->mc_fid.fid.ops = &rxm_mc_fi_ops;
+	rxm_mc->mc_fid.fi_addr = (uintptr_t) rxm_mc;
+
+	ofi_atomic_inc32(&av_set->ref);
+	rxm_mc->av_set = av_set;
+
+	return rxm_mc;
+}
+
 static int rxm_join_coll(struct fid_ep *ep, const void *addr, uint64_t flags,
-		    struct fid_mc **mc, void *context)
+			struct fid_mc **mc, void *context)
 {
+	struct fi_collective_addr *c_addr;
+	const struct fid_av_set *set;
+	struct util_av_set *av_set;
+	struct rxm_mc *rxm_mc;
 	struct rxm_ep *rxm_ep;
+	int ret;
 
 	if (!(flags & FI_COLLECTIVE))
 		return -FI_ENOSYS;
 
-	rxm_ep = container_of(ep, struct rxm_ep, util_ep.ep_fid);
+	c_addr = (struct fi_collective_addr *)addr;
+	set = c_addr->set;
+	av_set = container_of(set, struct util_av_set, av_set_fid);
 
-	return fi_join(rxm_ep->util_coll_ep, addr, flags, mc, context);
+	rxm_mc = rxm_create_mc(av_set, context);
+	if (!rxm_mc)
+		return -FI_ENOMEM;
+	rxm_mc->context = context;
+	rxm_ep = container_of(ep, struct rxm_ep, util_ep.ep_fid);
+	ret = fi_join(rxm_ep->util_coll_ep, addr, flags,
+			&rxm_mc->util_coll_mc_fid, rxm_mc);
+	if (ret) {
+		fi_close(&rxm_mc->mc_fid.fid);
+	} else if (rxm_ep->offload_coll_ep) {
+		ret = fi_join(rxm_ep->offload_coll_ep, addr, flags,
+			&rxm_mc->offload_coll_mc_fid, rxm_mc);
```

Comment:
You either need to serialize the join calls -- don't start the second join until the first completes -- or you need some sort of different context for the joins, so you can distinguish which one completed in the callback.

Reply:
fi_join() are serialized now